// node-ipinfo/dist/src/ipinfoLiteWrapper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var node_fetch_1 = __importDefault(require("node-fetch"));
var utils_1 = require("../config/utils");
var lruCache_1 = __importDefault(require("./cache/lruCache"));
var apiLimitError_1 = __importDefault(require("./errors/apiLimitError"));
var subnet_check_1 = require("subnet-check");
var common_1 = require("./common");
var version_1 = __importDefault(require("./version"));
var clientUserAgent = "IPinfoClient/nodejs/".concat(version_1.default);
var IPinfoLiteWrapper = (function () {
    function IPinfoLiteWrapper(token, cache, timeout, i18nData, baseUrl) {
        this.token = token;
        this.euCountries =
            (i18nData === null || i18nData === void 0 ? void 0 : i18nData.euCountries) && (i18nData === null || i18nData === void 0 ? void 0 : i18nData.euCountries.length) !== 0
                ? i18nData.euCountries
                : utils_1.defaultEuCountries;
        this.cache = cache ? cache : new lruCache_1.default();
        this.timeout =
            timeout === null || timeout === undefined
                ? common_1.REQUEST_TIMEOUT_DEFAULT
                : timeout;
        this.baseUrl = baseUrl || "https://".concat(common_1.HOST_LITE);
    }
    IPinfoLiteWrapper.cacheKey = function (k) {
        return "".concat(k, ":").concat(common_1.CACHE_VSN);
    };
    IPinfoLiteWrapper.prototype.fetchApi = function (path_1) {
        return __awaiter(this, arguments, void 0, function (path, init) {
            var headers, request, url;
            if (init === void 0) { init = {}; }
            return __generator(this, function (_a) {
                headers = {
                    Accept: "application/json",
                    Authorization: "Bearer ".concat(this.token),
                    "Content-Type": "application/json",
                    "User-Agent": clientUserAgent
                };
                request = Object.assign({
                    timeout: this.timeout,
                    method: "GET",
                    compress: false
                }, init, { headers: Object.assign(headers, init.headers) });
                url = [this.baseUrl, path].join(!this.baseUrl.endsWith("/") && !path.startsWith("/") ? "/" : "");
                return [2, (0, node_fetch_1.default)(url, request).then(function (response) {
                        if (response.status === 429) {
                            throw new apiLimitError_1.default();
                        }
                        if (response.status >= 400) {
                            throw new Error("Received an error from the IPinfo API " +
                                "(using authorization ".concat(headers["Authorization"], ") ") +
                                "".concat(response.status, " ").concat(response.statusText, " ").concat(response.url));
                        }
                        return response;
                    })];
            });
        });
    };
    IPinfoLiteWrapper.prototype.lookupIp = function () {
        return __awaiter(this, arguments, void 0, function (ip) {
            var data;
            var _this = this;
            if (ip === void 0) { ip = undefined; }
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (ip && this.isBogon(ip)) {
                            return [2, {
                                    ip: ip,
                                    bogon: true
                                }];
                        }
                        if (!ip) {
                            ip = "me";
                        }
                        return [4, this.cache.get(IPinfoLiteWrapper.cacheKey(ip))];
                    case 1:
                        data = _a.sent();
                        if (data) {
                            return [2, data];
                        }
                        return [2, this.fetchApi(ip).then(function (response) { return __awaiter(_this, void 0, void 0, function () {
                                var data, ipinfo;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4, response.json()];
                                        case 1:
                                            data = _a.sent();
                                            ipinfo = {
                                                ip: data.ip,
                                                asn: data.asn,
                                                asName: data.as_name,
                                                asDomain: data.as_domain,
                                                countryCode: data.country_code,
                                                country: data.country,
                                                continentCode: data.continent_code,
                                                continent: data.continent,
                                                isEU: this.euCountries.includes(data.country_code)
                                            };
                                            this.cache.set(IPinfoLiteWrapper.cacheKey(ip), ipinfo);
                                            return [2, ipinfo];
                                    }
                                });
                            }); })];
                }
            });
        });
    };
    IPinfoLiteWrapper.prototype.isBogon = function (ip) {
        if (ip != "") {
            for (var _i = 0, BOGON_NETWORKS_1 = common_1.BOGON_NETWORKS; _i < BOGON_NETWORKS_1.length; _i++) {
                var network = BOGON_NETWORKS_1[_i];
                if ((0, subnet_check_1.isInSubnet)(ip, network)) {
                    return true;
                }
            }
        }
        return false;
    };
    return IPinfoLiteWrapper;
}());
exports.default = IPinfoLiteWrapper;
